15: Redux I

gabrielizalo opened the pull request:
# Summary

* I moved the game from React states to the almighty Redux.


# Trello Card

https://trello.com/c/mxCofeK3/14-redux-i-actions-reducers-store-todo-finger


# Screenshot

<img width="1669" alt="react app - redux" src="https://user-images.githubusercontent.com/656353/47799021-70111280-dcf7-11e8-8c6a-5a1312705baa.png">

gabrielizalo committed:
Merge branch 'master' into 14-ReactApp-TicTacToe

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- I moved the game from React states to the almighty Redux.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- I moved the game from React states to the almighty Redux.

gabrielizalo committed:
Merge remote-tracking branch 'origin/15-Redux-I' into 15-Redux-I

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -7,47 +10,32 @@ import HistoryItem from './components/HistoryItem/';
-
   handleClick = i => {
-    const history = this.state.history.slice(0, this.state.stepNumber + 1);
+    const history = this.props.history.slice(0, this.props.stepNumber + 1);
```
You could use destructuring here

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -7,47 +10,32 @@ import HistoryItem from './components/HistoryItem/';
-      xIsNext: !this.state.xIsNext
-    });
+
+    squares[i] = this.props.xIsNext ? 'X' : 'O';
```
The same here

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -7,47 +10,32 @@ import HistoryItem from './components/HistoryItem/';
+
+    squares[i] = this.props.xIsNext ? 'X' : 'O';
+
+    this.props.makeMove(history, squares, history.length, this.props.xIsNext);
```
And here. I also think that you could reduce the amount of parameters that you send to the function because you could calculate the values of `history.length` and `xISNext` inside of the action

gabrielizalo replied:
You are right with history.length... But how with xIsNext? 😟

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -7,47 +10,32 @@ import HistoryItem from './components/HistoryItem/';
-      stepNumber: step,
-      xIsNext: step % 2 === 0
-    });
+    this.props.makeJump(step, this.props.xIsNext);
```
And here

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -7,47 +10,32 @@ import HistoryItem from './components/HistoryItem/';
   render() {
-    const history = this.state.history;
-    const current = history[this.state.stepNumber];
+    const history = this.props.history;
```
The same here

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -7,47 +10,32 @@ import HistoryItem from './components/HistoryItem/';
-    const history = this.state.history;
-    const current = history[this.state.stepNumber];
+    const history = this.props.history;
+    const current = history[this.props.stepNumber];
```
And here

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -7,47 +10,32 @@ import HistoryItem from './components/HistoryItem/';
     const winner = calculateWinner(current.squares);
 
-    const status = winner ? `Winner: ${winner}` : `Next player: ${this.state.xIsNext ? 'X' : 'O'}`;
+    const status = winner ? `Winner: ${winner}` : `Next player: ${this.props.xIsNext ? 'X' : 'O'}`;
```
And here

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -61,4 +49,27 @@ class Game extends Component {
 
-export default Game;
+Game.propTypes = {
+  history: PropTypes.arrayOf(PropTypes.objectOf).isRequired,
```
You should specify the types the `objectOf`

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -61,4 +49,27 @@ class Game extends Component {
+});
+
+const mapDispatchToProps = dispatch => ({
+  makeMove: (history, squares, stepNumber, xIsNext) =>
```
Check my recommendation above

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -61,4 +49,27 @@ class Game extends Component {
+const mapDispatchToProps = dispatch => ({
+  makeMove: (history, squares, stepNumber, xIsNext) =>
+    dispatch(makeMove(history, squares, stepNumber, xIsNext)),
+  makeJump: (stepNumber, xIsNext) => dispatch(makeJump(stepNumber, xIsNext))
```
The same here you could omit the `xIsNext` param

henryzarza commented on react/react-training/src/constants/index.js:
```diff
@@ -1 +1,2 @@
+// noinspection JSUnusedGlobalSymbols
```
Do you need this?

gabrielizalo replied:
Yes.. I need it so WebStorm don't do some recommendations.

henryzarza commented on react/react-training/src/index.js:
```diff
@@ -1,5 +1,13 @@
+import { store } from '@redux/store';
 
-ReactDOM.render(<Game />, document.getElementById('root'));
+const App = () => (
```
Isn't a good practice to use arrow function for the definition of components. Check this [style guide](https://github.com/Wolox/tech-guides/blob/master/react/style-guide.md#class-vs-reactcreateclass-vs-stateless)

henryzarza commented on react/react-training/src/redux/game/actions.js:
```diff
@@ -0,0 +1,26 @@
+// noinspection JSUnusedGlobalSymbols
```
Do you need this?

gabrielizalo replied:
Yesss... Or do you prefer I delete that comments?

henryzarza commented on react/react-training/src/redux/store.js:
```diff
@@ -0,0 +1,15 @@
+});
+
+/* eslint-disable no-underscore-dangle */
+// noinspection JSUnresolvedVariable
```
Do you need this?

henryzarza commented on react/react-training/src/utils/index.js:
```diff
@@ -1,5 +1,6 @@
 import { lines } from '@constants';
 
+// noinspection JSUnusedGlobalSymbols
```
The same question here

henryzarza commented on react/react-training/src/redux/game/reducer.js:
```diff
@@ -0,0 +1,27 @@
+  xIsNext: true
+};
+
+// noinspection JSUnusedGlobalSymbols
```
Do you need this?

henryzarza commented on react/react-training/src/redux/game/actions.js:
```diff
@@ -0,0 +1,26 @@
+// noinspection JSUnusedGlobalSymbols
+export const makeMove = (history, squares, stepNumber, xIsNext) => {
```
Move this function and `makeJump` inside of a object called `actionsCreators` and export it

henryzarza commented on react/react-training/src/redux/game/actions.js:
```diff
@@ -0,0 +1,26 @@
+  history = history.concat([{ squares }]);
+  xIsNext = !xIsNext;
+  return {
+    type: 'MAKE_MOVE',
```
Create a const called `actionTypes` and put in this value and `MAKE_JUMP `

henryzarza commented on react/react-training/src/redux/game/actions.js:
```diff
@@ -0,0 +1,26 @@
+    payload: {
+      history,
+      squares,
+      stepNumber,
```
You can calculate this value in the reducer

henryzarza commented on react/react-training/src/redux/game/actions.js:
```diff
@@ -0,0 +1,26 @@
+      history,
+      squares,
+      stepNumber,
+      xIsNext
```
The same here

henryzarza commented on react/react-training/src/redux/game/actions.js:
```diff
@@ -0,0 +1,26 @@
+
+// noinspection JSUnusedGlobalSymbols
+export const makeJump = (stepNumber, xIsNext) => {
+  xIsNext = stepNumber % 2 === 0;
```
The same here

henryzarza commented on react/react-training/src/redux/game/reducer.js:
```diff
@@ -0,0 +1,27 @@
+// noinspection JSUnusedGlobalSymbols
+export const reducer = (state = initialState, action) => {
+  switch (action.type) {
+    case 'MAKE_MOVE':
```
Use the `actionType` here

henryzarza commented on react/react-training/src/redux/game/reducer.js:
```diff
@@ -0,0 +1,27 @@
+};
+
+// noinspection JSUnusedGlobalSymbols
+export const reducer = (state = initialState, action) => {
```
Export as default

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fixed destructurings.
- Removed from makeMove the parameter stepNumber.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fixed PropType definition of History.
- Changed App arrow function to a Normal function.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- The redux actions now are part of "ActionsCreators" object.
- Deleted Webstorm comments.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Added Middleware, Compose and Redux Thunk to fix Actions.
- Removed from MakeMove the parameter xIsNext using dispatch into action.
-

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Renamed the reducer as GameReduceer and exported as default
- Created a new const ActionTypes as an enumerator in the Constants
- Removed the parameter xIsNext from makeJump action
-

gabrielizalo commented:
All recommendations are done 👍

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -61,4 +51,26 @@ class Game extends Component {
 
-export default Game;
+Game.propTypes = {
+  history: PropTypes.arrayOf(PropTypes.object).isRequired,
```
Change the `object` for `shape({VALUES})`

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -61,4 +51,26 @@ class Game extends Component {
+  makeJump: PropTypes.func.isRequired
+};
+
+const mapStateToProps = state => ({
```
You can use destructuring here

henryzarza commented on react/react-training/src/constants/index.js:
```diff
@@ -1 +1,5 @@
 export const lines = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], [1, 4, 7], [2, 5, 8], [0, 4, 8], [2, 4, 6]];
+export const actionTypes = {
```
This should be in `game/actions`

henryzarza commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -7,47 +10,34 @@ import HistoryItem from './components/HistoryItem/';
-      xIsNext: !this.state.xIsNext
-    });
+
+    squares[i] = xIsNext ? 'X' : 'O';
```
I think that you can calculate this value in `makeMove` action

gabrielizalo replied:
This correction will be ignored.

henryzarza reviewed (changes requested):
Very good, only a few change and we're ready

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fix the compose error because it was not imported.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Using destructuring.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fixing the type checking from object to the Shape PropType.

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Moving the actionType constant to actions in Redux.

gabrielizalo commented:
👌 
We are ready
🤘

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fixing the propTypes of history into Game.

henryzarza reviewed (dismissed):
Good!

wfolini commented on react/react-training/src/app/screens/Game/index.js:
```diff
@@ -61,4 +51,30 @@ class Game extends Component {
+  makeJump: stepNumber => dispatch(Actions.makeJump(stepNumber))
+});
+
+export default connect(
```
Split this component into a smart and dumb component. Check this post: https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0
https://medium.com/@learnreact/container-components-c0e67432e005

wfolini reviewed (changes requested):
Excellent redux refactor! 👏 👏  Just one comment!

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Now, the Game Component, is a Container and it uses a presentational
  component; GameViewer.

gabrielizalo commented:
We are readyyyy 👍

wfolini commented on react/react-training/src/app/screens/Game/components/GameViewer/index.js:
```diff
@@ -0,0 +1,51 @@
+import React, { Component } from 'react';
```
The refactor is fine! but isn't necessary to extract it to another component. Usually we extract the dumb component to a file `layout.js` and the smart to an `index.js` file, in the same component folder, in this case the `Game` folder and this file will be `layout.js`. 
You can check that here https://github.com/Wolox/tech-guides/blob/master/react/style-guide.md#folder-structure

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Transformed the 'GameViewer' component to 'layout.js'

gabrielizalo commented:
Listooo jejejeje 💯... ¡Ahora si que fue!

wfolini reviewed (dismissed):
👌

gabrielizalo committed:
Doing "ReactApp - Tic Tac Toe"
- Fixing names of Layout and Container in the game Component.
